packages/sqlite/src/lib/LibSqlClient.ts

```diff
@@ -0,0 +1,145 @@
+                values (?, ?, ?, ?, ?, ?, ?)`,
+          args: [
+            message.id || randomUUID(),
+            streamName,
+            category,
+            message.type,
+            message.data || null,
+            message.meta || null,
+            ++position,
+          ],
+        })
+      }
+      await trx.commit()
+    } catch (err: any) {
+      await trx.rollback()
+      throw err
+    }
+
+    return { type: "Written", position }
+  }
+}
+
+export class LibSqlReader {
+  constructor(private readonly client: Client) {}
+
+  async readLastEventWithType(streamName: string, eventType: string) {
+    const result = await this.client.execute({
+      sql: "select * from messages where stream_name = ? and type = ? order by position desc limit 1",
+      args: [streamName, eventType],
+    })
+    if (result.rows.length === 0) return null
+    return Parse.row(result.rows[0])
+  }
+
+  async readLastEvent(streamName: string, eventType?: string) {
+    if (eventType != null) return this.readLastEventWithType(streamName, eventType)
+    const result = await this.client.execute({
+      sql: "select * from messages where stream_name = ? order by position desc limit 1",
+      args: [streamName],
+    })
+    if (result.rows.length === 0) return null
+    return Parse.row(result.rows[0])
+  }
+
+  async readStream(
+    streamName: string,
+    fromPosition: bigint,
+    batchSize: number,
+  ): Promise<ITimelineEvent<Format>[]> {
+    const result = await this.client.execute({
+      sql: "select * from messages where stream_name = ? and position >= ? order by position limit ?",
+      args: [streamName, fromPosition.toString(), batchSize],
+    })
+    return result.rows.map(Parse.row)
+  }
+}
+
+type DbRow = {
+  id: string
+  time: string
+  type: string
+  data?: string
+  meta?: string
+  position: string
+  global_position: string
+}
+
+namespace Parse {
+  export const string = (s: unknown): string => {
+    if (typeof s === "string") return s
+    throw new Error(`Expected string, got ${typeof s}`)
+  }
+  export const bigint = (s: unknown): bigint => {
+    switch (typeof s) {
+      case "string":
+      case "number":
+        return BigInt(s)
+      default:
+        throw new Error(`Expected string, got ${typeof s}`)
+    }
+  }
+
+  export const date = (s: unknown): Date => {
+    if (typeof s === "string") return new Date(s)
+    throw new Error(`Expected string, got ${typeof s}`)
+  }
+
+  export const nullable =
+    <T>(parser: (s: unknown) => T) =>
+    (s: unknown): T | undefined => {
+      if (s == null) return
+      return parser(s)
+    }
+
+  export const row = (r: any): ITimelineEvent<Format> => {
+    const row = r as DbRow
+    const data = nullable(string)(row.data)
+    const meta = nullable(string)(row.meta)
+    const type = string(row.type)
+    const time = date(row.time)
+    const index = bigint(row.position)
```

Comment:
indx would align so pretty

Reply:
you're serious aren't you?

Reply:
I've done that in the past and partially recovered my sanity, but have not forgotten the good times!